### DISCUSSION TITLE
Start DAGs based on async event on queues

### DISCUSSION THREAD

**Discussion by MattiaGallegati** [apache/airflow#31119] [Q&A]
Hello,

Let's suppose that a user wants to start an Airflow DAG each time a message reaches a queue or topic on an event broker or a streaming broker (e.g. AMQ, Kafka, ecc..), is there a ""standard"" way of doing this in Airflow?
I explored Airflow Sensor, btw, from what I understood, they work in a quite different way as they need to be created as task of specific DAG and the DAG needs to be "started" based on schedule interval or other mechanism. Sensor are simple tasks that stay in "polling" state until they receive the "kick". Once the DAG has been started and the Sensor receive its "kick" the DAG (and the sensor) will not run again until the DAG's next schedule date.
Am i right?
We can see new solutions (e.g. Dagster) that manage Sensors as first citizien of the platform, they are indipendent, constantly running and they can be fully customized. The Sensors start new runs every time the condition is met.
Do we have something similar in Airflow? Can it be considered a new potential feature for the platform?

Thank you

**Comment by potiuk** [comment]
It is somethig that will come and likely quite soon as a way to trigger DAGs based on some dataset change events. We are going to leverage data-aware scheduling: https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/datasets.html  - triggering a DAG based on a change in a dataset that you define as an input is somethig that will be implemented in one of the next major releases most likely. It's been considered as being part of the original implementation: AIP-48: https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-48+Data+Dependency+Management+and+Data+Driven+Scheduling but it has been decided to be implemented as follow-up, we considered cross-dag-dependencies using dataset a much more important feature.

We are going to - most likely  - use the infrastructure we already implemented for another feature - Deferrable OPerators and Triggers https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/deferring.html#using-deferrable-operators - where AsyncIO Python functionality is used to be able to handle external even triggering using very efficient Async IO triggers. One of the problems with such a "polling" mechanism if done in a trivial, synchronous way is that it takes a lot of resources to just poll for possible multiple external triggers, but with async io and scalable Triggerer running asyncio loop, we can  efficiently subscribe to many thousands of such event sources with single async io loop and do such polling efficiently.

This has also been `deferred` (pun intended) for later in AIP-44 where Deferrable/Async operators have been implemented  https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=177050929 

This is not a highest priority for us though, especially that you can very easily implement something like that on your own in a way that is native to the environment you are in. For example all the major cloud providers have mechanism that allow you to - very easily - trigger a webhook and post a "Trigger DAG" API call for Airlfow: You have a Python and GO client to trigger the dag externally, as well as REST API that they use: https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html#operation/post_dag_run and writing an AWS  lambda function for example or GCP pubsub trigger a DAG is generally a trivial effort, if you run it on premises, you can also usually very easily extend whatever is going to be the source of such dag triggering to add a POST message to trigger the DAG in Airlfow from there - and until we implement it in a scalable, generalizable way - this is the recommended solution.

**Comment by zoid-w** [comment]
It might be worth taking a look at : https://airflow.apache.org/docs/apache-airflow-providers-apache-kafka/stable/sensors.html. In particular https://github.com/apache/airflow/blob/providers-apache-kafka/1.1.0/tests/system/providers/apache/kafka/example_dag_event_listener.py is quite close to what you have in mind and I have implemented similar use cases. It would indeed be awesome if we could the dataset capabilities to achieve this in the future and not rely on the triggerdagRun operator. Also the recent added continuous scheduling interval might help overcome your 'scheduling issue'. Note that if you are implementing some similar logic ensure your sensor is a deferable one to prevent occupying a slot on your executor continuously.
